### PR TITLE
Unified Downloader

### PR DESCRIPTION
**Pull request by gunner272**
Unified Downloader's primary aim is to remove the downloading of light curve files from the Light Curve module.
This is done with help of custom clients written for LC sources.In addition VSO module is also made compatible with this. 
The downloader has same query based capability as the vso module.
Unified Downloader has a factory based implementation requiring the client to have validation class method '_can_handle_query'. 

PS: This is not pep8 compliant yet in various parts.I am working on that

---------------------------------------

ToDo:

* [x] A Rebase onto master.
* [ ] Testing to make sure that this plays nice with EVE and similar where both the VSO and the specialised client return results.
* [ ] Guide documentation on how this affects Lightcurve.
* [ ] Integration with JSOC. (@Cadair)
* [ ] Try and unify the common attributes of `UnifiedResponse`
* [ ] Test coverage?!
* [ ] Full, detailed code review.

### CODE REVIEW

**Comment by aringlis** [comment]
@Cadair where are we with this? I get a couple of import errors when I try to use unified downloader. I pulled Rishabh's branch for testing purposes.

1) ImportError: cannot import name FileDownloader

Looks like a simple fix - FileDownloader appears to have changed name to just Downloader.

2) AttributeError: type object 'VSOClient' has no attribute '_can_handle_query'

I presume this function is used when deciding which of the possible clients to use to service a UniDown query.

**Comment by gunner272** [comment]
@aringlis: try now should work

**Comment by aringlis** [comment]
Ok, great - those errors I mentioned were fixed!

**Comment by aringlis** [comment]
So there's a print bug when trying to display the results of a UnifiedDownloader query that ends up being serviced by the VSO, e.g.

    results = UnifiedDownloader.query(attrs.Time("2012/1/1", "2012/1/2"),
                               attrs.Instrument('eve'),attrs.Detector('esp'))
    print results
    AttributeError: Time instance has no attribute 't1'



I think @gunner272 @wafels you're already aware of this - is there a handy fix?

**Comment by Cadair** [comment]
replace `.t1` with `.start`

**Comment by gunner272** [comment]
@Cadair, that doesnt seem to work
Time attribute creates start and end objects via calling parse_time which was returning a string,
I think we need to change that to TimeRange

**Comment by Cadair** [comment]
oh, yeah a `TimeRange` would be better.

**Comment by wafels** [comment]
It seems that the VSO query response Time object does not parse the times
it gets from the VSO.  Therefore the time properties of the VSO query
response object are not SunPy times.  Is it better to have these times as
SunPy times?

**Comment by Cadair** [comment]
as long as it dosen't break anything.

**Comment by aringlis** [comment]
Depending on the client that's actually used (e.g. EVEClient, LYRAClient, VSO), the UnifiedResponse is not necessarily unified! For example, a response generated from the EVEClient has .url attribute, but a response generated from the VSO does not. Also, .time attributes returned from a VSO query are different, as mentioned above.

Do we need to unify this more, and if so, how should we do it?

**Comment by Cadair** [comment]
I think there will be some differences, the `JSOCClient` is quite different for example. The common attributes such as `.time` should be the same though.

File: sunpy/net/unifieddownloader/sources/lyra.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	----------
+	date : datetime 
+
+        Returns
+	-------
+	string representing URL
+	"""
+        if not isinstance(date, datetime.date):
+            raise ValueError("This method requires a date")
+        filename = "lyra_%s000000_lev%d_%s.fits" % (date.strftime('%Y%m%d-'), 2, 'std')
```
Comment (dpshelio) [inline sunpy/net/unifieddownloader/sources/lyra.py]: This should be updated to the new way of formating strings:
```python
filename = "lyra_{0:%y%m%s}_lev{1}_{2}.fits".format(date, '2', 'std')
```

**Comment by Cadair** [comment]
I moved this into a feature branch to facilitate everyone working on it for the sprint at the end of next week.

**Pull request closed without merging**